photon-client/src/main/scala/com/linkedin/photon/ml/hyperparameter/ShrinkSearchRange.scala

```diff
@@ -0,0 +1,126 @@
+                priorJsonString: String,
+                priorDefault: Map[String, String],
+                radius: Double,
+                seed: Long = System.currentTimeMillis): (DenseVector[Double], DenseVector[Double]) = {
+
+    val hyperparameterList = hyperParams.names
+    val ranges = hyperParams.ranges
+    val discreteParams = hyperParams.discreteParams
+    val numParams = ranges.length
+
+    // Get a [[Seq]] of (vectorized hyper-parameters, evaluationValue) tuples
+    val hyperparameterPairs = HyperparameterSerialization.priorFromJson(priorJsonString, priorDefault, hyperparameterList)
+
+    // Rescale the hyperparameters to [0,1]
+    val hyperparameterRescaled = VectorRescaling.rescalePriors(hyperparameterPairs, hyperParams)
+
+    // Combine hyperparameters as a dense matrix and evaluation value as a dense vector
+    val (overallPoints, overallEvals) = hyperparameterRescaled.map(x => (x._1.asDenseMatrix, DenseVector(x._2))).reduce(
+      (a, b) => (DenseMatrix.vertcat(a._1, b._1), DenseVector.vertcat(a._2, b._2))
+    )
+
+    // Fit Gaussian process regression model
+    val estimator = new GaussianProcessEstimator(kernel = new Matern52)
+    val model = estimator.fit(overallPoints, overallEvals)
+
+    // Sobol generator
+    val paramDistributions = {
+      val sobol = new SobolSequenceGenerator(numParams)
+      sobol.skipTo((seed % (Int.MaxValue.toLong + 1)).toInt)
+      sobol
+    }
+
+    // Draw candidates from a Sobol generator
+    val candidates = (1 until 1000).foldLeft(DenseMatrix(paramDistributions.nextVector)) { case (acc, _) =>
+      DenseMatrix.vertcat(acc, DenseMatrix(paramDistributions.nextVector))
+    }
+
+    // Select the best candidate
+    val predictions = model.predict(candidates)
+    val bestCandidate = selectBestCandidate(candidates, predictions._1)
+
+    // compute lower bound and upper bound
+    val upperBound = VectorRescaling.scaleBackward(
+      discretizeCandidate(bestCandidate.map(x => min(x + radius, 1.0)), discreteParams),
+      ranges,
+      discreteParams.keySet)
+    val lowerBound = VectorRescaling.scaleBackward(
+      discretizeCandidate(bestCandidate.map(x => max(x - radius, 0.0)), discreteParams),
+      ranges,
+      discreteParams.keySet)
+
+    discreteParams.foreach{
+      case (index, _) =>
+        upperBound(index) = min(upperBound(index) + 1, ranges(index).end)
+        lowerBound(index) = max(lowerBound(index) - 1, ranges(index).start)
+    }
+
+    (lowerBound, upperBound)
+  }
+
+  /**
+    * Selects the best candidate according to the predicted values, where "best" is defined as the largest
+    *
+    * @param candidates matrix of candidates
+    * @param predictions predicted values for each candidate
+    * @return the candidate with the best value
+    */
+  private def selectBestCandidate(candidates: DenseMatrix[Double],
+                                  predictions: DenseVector[Double]): DenseVector[Double] = {
+
```

Comment:
Is there a reason to copy and paste this function from `GaussianProcessSearch`?

Reply:
selectBestCandidate uses betterThan method for evaluator, which we cannot use without specifying the evaluator. This function will be re-written in atlas-ml so that we can directly call it.